feat: add tests for prebuild metrics

SasSwart opened the pull request:
This PR introduces tests for the prebuild system's metrics collector.

SasSwart committed:
finalise prebuild counter metric tests

Kira-Pilot committed:
fix: select default org in template form if only one exists (#16639)

resolves #16849 https://github.com/coder/internal/issues/147
![Screenshot 2025-02-19 at 9 06 16
PM](https://github.com/user-attachments/assets/2973d81d-7a74-4c82-aa6b-16d4a41eeb9a)

---------

Co-authored-by: ケイラ <mckayla@hey.com>

brettkolodny committed:
feat: add a paginated organization members endpoint (#16835)

Closes
[coder/internal#460](https://github.com/coder/internal/issues/460)

matifali committed:
chore: enable SBOM attestation for image builds (#16852)

- Added SBOM (Software Bill of Materials) generation during Docker build
to enhance traceability. Refer to Docker documentation on SBOM:
https://docs.docker.com/build/metadata/attestations/sbom/
- Updated Docker build scripts to use BuildKit for provenance and SBOM
support: https://docs.docker.com/build/metadata/attestations/
- Configured Docker daemon in dogfood image to support the Containerd
snapshotter feature to improve performance:
https://docs.docker.com/engine/storage/containerd/

> [!Important]
> We also need to enable `containerd` on depot runners.
> <img width="587" alt="image"
src="https://github.com/user-attachments/assets/1d7f87c7-fdcc-462a-babe-87ac6486ad09"
/>



## Testing

- Tested locally with ` docker buildx build --sbom=true --output
type=local,dest=out -f Dockerfile .` to verify that an SBOM file is
generated.
- Tested in
[CI](https://github.com/coder/coder/actions/runs/13731162662/job/38408790980?pr=16852#step:17:1)
to ensure the image builds without any errors.


Also closes coder/internal#88

matifali committed:
revert: "chore: enable SBOM attestation for image builds" (#16868)

Reverts coder/coder#16852

The CI failed to create the multi-arch manifest.

https://github.com/coder/coder/actions/runs/13773079355/job/38516182819#step:18:341

I personally think we should move to a [multi-arch
Dockerfile](https://docs.docker.com/build/building/multi-platform/#cross-compilation)
instead of creating the manifest manually.

EdwardAngert committed:
docs: convert alerts to use GitHub Flavored Markdown (GFM) (#16850)

followup to #16761 

thanks @lucasmelin !

+ thanks: @ethanndickson @Parkreiner @matifali @aqandrew 

- [x] update snippet
- [x] find/replace
- [x] spot-check


[preview](https://coder.com/docs/@16761-gfm-callouts/admin/templates/managing-templates/schedule)
(and others)

---------

Co-authored-by: EdwardAngert <17991901+EdwardAngert@users.noreply.github.com>
Co-authored-by: M Atif Ali <atif@coder.com>

jaaydenh committed:
fix: use correct permissions for CRUD of custom roles (#16854)

resolves coder/internal#428

The goal of the PR is to start using updateOrgRoles and deleteOrgRoles
permissions to gate custom roles functionality

```
		updateOrgRoles: {
			object: {
				resource_type: "assign_org_role",
				organization_id: organizationId,
			},
			action: "update",
		},
		deleteOrgRoles: {
			object: {
				resource_type: "assign_org_role",
				organization_id: organizationId,
			},
			action: "delete",
		}
```

SasSwart committed:
assert for label values in prebuild metrics tests

SasSwart committed:
add tests for prebuild gauge metrics

make -B fmt

mtojek commented:
I see it is marked as a draft, is it ready for review @SasSwart?

dannykopping commented:
Let's not worry too much about reviewing this PR; this is just going into the feature branch.
We'll review everything in detail once we start splitting things out.

dannykopping commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -10,30 +10,33 @@ import (
 var (
 	CreatedPrebuildsDesc   = prometheus.NewDesc("coderd_prebuilds_created", "The number of prebuilds created.", []string{"template_name", "preset_name"}, nil)
 	FailedPrebuildsDesc    = prometheus.NewDesc("coderd_prebuilds_failed", "The number of prebuilds that failed.", []string{"template_name", "preset_name"}, nil)
 	AssignedPrebuildsDesc  = prometheus.NewDesc("coderd_prebuilds_assigned", "The number of prebuilds that were assigned to a runner.", []string{"template_name", "preset_name"}, nil)
```
"to a runner"?

dannykopping commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -10,30 +10,33 @@ import (
+	"github.com/coder/coder/v2/coderd/prebuilds"
 )
 
 var (
```
The descriptions for failed, exhausted, actual, and desired are too vague; please add some more detail.

dannykopping replied:
👌 perfect, thanks

dannykopping commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+									TemplateID: template.ID,
+								})
+								require.NoError(t, err)
+								require.Equal(t, 1, len(templateVersions))
+
+								presets, err := db.GetPresetsByTemplateVersionID(ctx, templateVersions[0].ID)
+								require.NoError(t, err)
+								require.Equal(t, 1, len(presets))
```
I worry about just testing with 1 of each. We've seen this hide weird behaviour in other places.
If it's not too much trouble, can you expand this to include more - even if only 1 more?

dannykopping commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+
+								for _, preset := range presets {
+									checks := []metricCheck{
+										{"coderd_prebuilds_created", test.shouldIncrementPrebuildsCreated, true},
```
Nit: can you add the labels here? It'd aid readability.

SasSwart replied:
I've moved these up into the test case definitions. Please check now?

Its not quite what you had in mind, but I do think its the better structure on balance of consideration.

dannykopping replied:
My point was more that the reader then had to find and memorise the order of fields in `metricCheck`, but this is cool for now.

dannykopping commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+
+func findMetric(metricsFamilies []*prometheus_client.MetricFamily, name string, labels map[string]string) *prometheus_client.Metric {
+	for _, metricFamily := range metricsFamilies {
+		if metricFamily.GetName() == name {
```
Nit: keep happy path unintended; i.e. continue if the name doesn't match

mtojek replied:
+1

dannykopping commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+										}
+
+										require.NotNil(t, metric, "metric %s should exist", check.name)
+										if check.isCounter {
```
I _think_ this is where `promtestutil.ToFloat64` can help simplify things.

SasSwart replied:
Tried it. Didn't work because the metric here doesn't implement the collector interface. I checked the function documentation and they discourage using it pretty heavily. Going to stick to what I have for now if thats okay?

dannykopping reviewed (approved):
LGTM! Awesome.

mtojek commented:
> Let's not worry too much about reviewing this PR; this is just going into the feature branch.
> We'll review everything in detail once we start splitting things out.

I think the opposite way. Let's start looking into the code now, even giving a partial look/review, and make the team members familiar with it.

mtojek commented on coderd/database/queries/prebuilds.sql:
```diff
@@ -106,41 +107,14 @@ RETURNING *;
-GROUP BY t.name, tvp.name;
+    COUNT(*) as created_count,
+    COUNT(*) FILTER (WHERE pj.job_status = 'failed'::provisioner_job_status) as failed_count,
+    COUNT(*) FILTER (WHERE w.owner_id != 'c42fdf75-3097-471c-8c33-fb52454d81c0'::uuid) as claimed_count
```
I understand that this is a special user `prebuilds`? I'm not convinced about hardcoding it in the SQL query. Do you think we can pass from the backend?

SasSwart replied:
To clarify: Is this an instruction or are you asking for my opinion?

mtojek replied:
Always opinion, you're the owner of this code.

PS. I think Cian has also raised question about hardcoding ID.

mtojek commented on enterprise/coderd/coderd.go:
```diff
@@ -884,9 +883,9 @@ func (api *API) updateEntitlements(ctx context.Context) error {
 			reconciler, claimer, metrics := api.setupPrebuilds(enabled)
 			if api.PrebuildsReconciler != nil {
-				stopCtx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, errors.New("gave up waiting for reconciler to stop"))
+				stopCtx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, xerrors.New("gave up waiting for reconciler to stop"))
```
👍

johnstcn commented on coderd/database/queries/prebuilds.sql:
```diff
@@ -106,41 +107,14 @@ RETURNING *;
-GROUP BY t.name, tvp.name;
+    COUNT(*) as created_count,
+    COUNT(*) FILTER (WHERE pj.job_status = 'failed'::provisioner_job_status) as failed_count,
+    COUNT(*) FILTER (WHERE w.owner_id != 'c42fdf75-3097-471c-8c33-fb52454d81c0'::uuid) as claimed_count
```
I know this is pre-existing, but can you add a comment regarding this magic UUID?

mtojek commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -10,30 +10,33 @@ import (
 	FailedPrebuildsDesc    = prometheus.NewDesc("coderd_prebuilds_failed", "The number of prebuilds that failed.", []string{"template_name", "preset_name"}, nil)
 	AssignedPrebuildsDesc  = prometheus.NewDesc("coderd_prebuilds_assigned", "The number of prebuilds that were assigned to a runner.", []string{"template_name", "preset_name"}, nil)
-	UsedPresetsDesc        = prometheus.NewDesc("coderd_presets_used", "The number of times a preset was used.", []string{"template_name", "preset_name"}, nil)
+	UsedPresetsDesc        = prometheus.NewDesc("coderd_prebuilds_used_presets", "The number of times a preset was used to build a prebuild.", []string{"template_name", "preset_name"}, nil)
```
for later: you may want to review metric names with Spike (quality proved in the past :)

mtojek commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -53,19 +56,42 @@ func (mc *MetricsCollector) Collect(metricsCh chan<- prometheus.Metric) {
 	defer cancel()
 	// nolint:gocritic // just until we get back to this
-	metrics, err := mc.database.GetPrebuildMetrics(dbauthz.AsSystemRestricted(ctx))
+	prebuildCounters, err := mc.database.GetPrebuildMetrics(dbauthz.AsSystemRestricted(ctx))
```
Hm... just counters? no gauges in the future?

mtojek commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -10,30 +10,33 @@ import (
 )
 
 var (
 	CreatedPrebuildsDesc   = prometheus.NewDesc("coderd_prebuilds_created", "The number of prebuilds created.", []string{"template_name", "preset_name"}, nil)
```
is there a reason why `desc` are exposed?

SasSwart replied:
nope. fixed

mtojek commented on enterprise/coderd/prebuilds/metricscollector.go:
```diff
@@ -53,19 +56,42 @@ func (mc *MetricsCollector) Collect(metricsCh chan<- prometheus.Metric) {
+
+	for _, preset := range state.Presets {
+		if !preset.UsingActiveVersion {
+			continue
```
Perhaps add debug or trace here

mtojek commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+			shouldSetEligiblePrebuilds:       ptr.To(0.0),
+		},
+	}
+	for _, test := range tests {
+		for _, transition := range test.transitions {
+			for _, jobStatus := range test.jobStatuses {
+				for _, initiatorID := range test.initiatorIDs {
+					for _, ownerID := range test.ownerIDs {
+						t.Run(fmt.Sprintf("transition:%s/jobStatus:%s", transition, jobStatus), func(t *testing.T) {
```
oh man.. what a complexity :D 

Perhaps take a look at `insights-metrics.json`. You can define test results as .json files, and have them unmarshalled and compared.

SasSwart replied:
what specifically is the complexity that you are referring to?

johnstcn commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+
+										require.NotNil(t, metric, "metric %s should exist", check.name)
+										if check.isCounter {
+											require.Equal(t, *check.value, metric.GetCounter().GetValue(), "counter %s value mismatch", check.name)
```
suggestion: this might be better as an `assert`; failing the test immediately might hide additional clues from subsequent failing assertions.

johnstcn commented on enterprise/coderd/prebuilds/metricscollector_test.go:
```diff
@@ -1,73 +1,247 @@
+		jobStatuses                      []database.ProvisionerJobStatus
+		initiatorIDs                     []uuid.UUID
+		ownerIDs                         []uuid.UUID
+		shouldIncrementPrebuildsCreated  *float64
+		shouldIncrementPrebuildsFailed   *float64
+		shouldIncrementPrebuildsAssigned *float64
+		shouldSetDesiredPrebuilds        *float64
+		shouldSetActualPrebuilds         *float64
+		shouldSetEligiblePrebuilds       *float64
```
Would it not be clearer to allow each test case to set a `map[string]float64` of the expected metric names / deltas?

johnstcn commented on enterprise/coderd/coderd.go:
```diff
@@ -884,9 +883,9 @@ func (api *API) updateEntitlements(ctx context.Context) error {
 			reconciler, claimer, metrics := api.setupPrebuilds(enabled)
 			if api.PrebuildsReconciler != nil {
-				stopCtx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, errors.New("gave up waiting for reconciler to stop"))
+				stopCtx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, xerrors.New("gave up waiting for reconciler to stop"))
```
Also out of scope of this PR: stopping the prebuilds reconciler on-the-fly due to a licensing change is a pretty big deal. I assume it'll have to scale down all of the pools and delete all of the resources. This will be a pretty expensive operation and potentially disruptive in the case of an operator accidentally forgetting to insert an updated license.

dannykopping replied:
> I assume it'll have to scale down all of the pools and delete all of the resources.

That won't be the case. Stopping the reconciler just stops the reconciliation loop, and does not affect any of the running prebuilds.

We _could_ implement that, but the likelihood of someone removing their premium license is low. We'll punt for now. What we've implemented so far is just the ability to handle license changes (mostly for tests).

mtojek reviewed (commented):
@SasSwart should we switch the PR destination to `dk/prebuilds`?

johnstcn commented on enterprise/coderd/coderd.go:
```diff
@@ -674,7 +673,7 @@ func (api *API) Close() error {
 
 	if api.PrebuildsReconciler != nil {
-		ctx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, errors.New("gave up waiting for reconciler to stop"))
+		ctx, giveUp := context.WithTimeoutCause(context.Background(), time.Second*30, xerrors.New("gave up waiting for reconciler to stop"))
```
out of scope of this PR: I'd like to see the timeout logic moved inside the prebuilds package; this seems to be the norm for other components.

SasSwart committed:
Merge branch 'dk/quotas' into jjs/prebuilds-metrics

dannykopping committed:
Extraneous prebuilds can ONLY be running, so if one is already deleting it cannot be extraneous

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Prevent duplicate preset names in a template

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

mafredri committed:
feat(agent): set additional login vars, LOGNAME and SHELL (#16874)

This change stes additional env vars. This is useful for programs that
assume their presence (for instance, Zed remote relies on SHELL).

See `man login`.

SasSwart committed:
make prebuild metric descriptors private

dannykopping committed:
Add more log context

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

SasSwart committed:
provide more specific descriptions for prebuild metrics

dannykopping commented:
> @SasSwart should we switch the PR destination to `dk/prebuilds`?

No, this is first going into `dk/quotas` which we're accumulating some additional changes in first before merging into `dk/prebuilds`.

ThomasK33 committed:
fix(flake.nix): synchronize playwright version in nix and package.json (#16715)

Ensure that the version of Playwright installed with the Nix flake is
equal to the one specified in `site/package.json.` -- This assertion
ensures that `pnpm playwright:install` will not attempt to download
newer browser versions not present in the Nix image, fixing the startup
script and reducing the startup time, as `pnpm playwright:install` will
not download or install anything.

We also pre-install the required Playwright web browsers in the dogfood
Dockerfile. This change prevents us from redownloading system
dependencies and Google Chrome each time a workspace starts.

Change-Id: I8cc78e842f7d0b1d2a90a4517a186a03636c5559
Signed-off-by: Thomas Kosiewski <tk@coder.com>

Signed-off-by: Thomas Kosiewski <tk@coder.com>

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into dk/prebuilds

SasSwart committed:
improve test structure

SasSwart committed:
make -B fmt

SasSwart committed:
rename variable

SasSwart committed:
make the linter happy

SasSwart committed:
make the linter happy

SasSwart committed:
Merge remote-tracking branch 'origin/dk/prebuilds' into jjs/prebuilds-metrics

SasSwart committed:
fix tests

SasSwart committed:
make -B gen

SasSwart committed:
add prebuild mutual exclusion test

SasSwart committed:
reduce log level to get tests to pass

SasSwart committed:
fix tests

dannykopping committed:
Implement missing dbauthz tests

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Add migration fixtures for prebuilds

Also removed schedule table; we are not using it yet

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Accounting for dbmem

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

SasSwart commented:
This work has been merged into `dk/prebuilds`. Formal review will continue in a separate PR as part of a larger effort to reintegrate prebuilds into mainline. cc @dannykopping cc @mtojek

mtojek commented:
👍
